cmd/proxy/actions/backgroundsync.go

```diff
@@ -0,0 +1,113 @@
+package actions
+
+import (
+	"errors"
+	"io/ioutil"
+	"time"
+
+	"github.com/gobuffalo/buffalo/worker"
+	"github.com/gobuffalo/envy"
+	"github.com/gomods/athens/pkg/eventlog"
+	"github.com/gomods/athens/pkg/eventlog/olympus"
+	proxystate "github.com/gomods/athens/pkg/proxy/state"
+	"github.com/gomods/athens/pkg/storage"
+	olympusStore "github.com/gomods/athens/pkg/storage/olympus"
+)
+
+// GetProcessModuleJob porcesses job from a queue and downloads missing module
+func GetProcessModuleJob(s storage.Backend, ps proxystate.Store) worker.Handler {
+	return func(args worker.Args) error {
+		olympusEndpoint, ok := args[workerEndpointKey].(string)
+		if !ok {
+			return errors.New("olympus endpoint not provided")
+		}
+		event, ok := args[workerEventKey].(eventlog.Event)
+		if !ok {
+			return errors.New("event to process not provided")
+		}
+
+		if s.Exists(event.Module, event.Version) {
+			return nil
+		}
+
+		// get module info
+		os := olympusStore.NewStorage(olympusEndpoint)
+		version, err := os.Get(event.Module, event.Version)
+		if err != nil {
+			return err
+		}
+
+		zip, err := ioutil.ReadAll(version.Zip)
+		if err != nil {
+			return err
+		}
+
+		return s.Save(event.Module, event.Version, version.Mod, zip)
+	}
+}
+
+// SyncLoop is synchronization background job meant to run in a goroutine
+// pulling event log from olympus
+func SyncLoop(s storage.Backend, ps proxystate.Store, w worker.Worker) {
+	olympusEndpoint, sequenceID := getLoopState(ps)
+
+	for {
+		select {
+		case <-time.After(30 * time.Second):
+			ee, err := getEventLog(olympusEndpoint, sequenceID)
+
+			if err != nil {
+				ps.Clear()
```

Comment:
think what about jobs in queue, you cannot process them as they point to old olympus

Reply:
you're not even following redirects from global Olympus to local instance, need to change the configuration on redirect.
on redirect return error with new uri,
set the uri to the state
next batch will be fetched from new location

Reply:
f: added variable which holds current endpoint, jobs check associated==current